### PR TITLE
TMI2-446: Spotlight token refresh

### PR DESCRIPTION
**Pull request by iaincooper-tco**
## Description

https://technologyprogramme.atlassian.net/browse/TMI2-446

Scheduled job to refresh the Spotlight Access token

## Type of change

Please check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [X] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [X] Unit Test
- [ ] Integration Test (if applicable)
- [ ] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

# Checklist:

- [ ] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [ ] I have performed a self-review of my code.
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

### CODE REVIEW

File: src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java [hunk 1]
```diff
@@ -125,16 +129,62 @@ public void exchangeAuthorizationToken(String authorizationCode, String state) t
             log.debug("new Access token: {}", accessToken);
             log.debug("new Refresh token: {}", refreshToken);
 
-            this.updateSecret("access_token", accessToken);
-            this.updateSecret("refresh_token", refreshToken);
+            this.updateSecret(ACCESS_TOKEN_NAME, accessToken);
+            this.updateSecret(REFRESH_TOKEN_NAME, refreshToken);
 
         } catch (IOException e) {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java:135]: Can do this as both exceptions are handled in the same way:

`catch(IOException | JSONException e)`

File: src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java [hunk 2]
```diff
@@ -125,16 +129,62 @@ public void exchangeAuthorizationToken(String authorizationCode, String state) t
...
+            log.debug("responseJSON: {}", responseJSON);
+
+            String accessToken = responseJSON
+                    .get("access_token")
+                    .toString();
+
+            log.debug("new Access token: {}", accessToken);
+
+            this.updateSecret(ACCESS_TOKEN_NAME, accessToken);
+        } catch (IOException e) {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java:178]: same here `catch(IOException | JSONException e)`

File: src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java [hunk 3]
```diff
@@ -43,6 +44,9 @@ public class SpotlightService {
     private String codeVerifier;
     private String codeChallenge;
 
+    private final String ACCESS_TOKEN_NAME = "access_token";
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java:47]: these can be static:

`    private static final String ACCESS_TOKEN_NAME = "access_token";
     private static final String REFRESH_TOKEN_NAME = "refresh_token";
`

File: src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java [hunk 4]
```diff
@@ -119,4 +119,44 @@ public RedirectView callback(@RequestParam String code, @RequestParam String sta
     }
 
 
+    @GetMapping("/oauth/refresh")
+    public ResponseEntity refresh(final HttpServletRequest httpRequest) throws Exception {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java:123]: Can add return type here:

`ResponseEntity<Void>`

File: src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java [hunk 5]
```diff
@@ -119,4 +119,44 @@ public RedirectView callback(@RequestParam String code, @RequestParam String sta
...
+
+            SpotlightOAuthAudit spotlightOAuthAudit = SpotlightOAuthAudit.builder()
+                    .user(user)
+                    .event(SpotlightOAuthAuditEvent.REFRESH)
+                    .status(SpotlightOAuthAuditStatus.FAILURE)
+                    .build();
+
+            spotlightService.saveAudit(spotlightOAuthAudit);
+
+            throw new Exception("Error refreshing Spotlight authorization token");
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java:157]: I think a dedicated exception might be better here?

File: src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java [hunk 6]
```diff
@@ -119,4 +119,44 @@ public RedirectView callback(@RequestParam String code, @RequestParam String sta
     }
 
 
+    @GetMapping("/oauth/refresh")
+    public ResponseEntity refresh(final HttpServletRequest httpRequest) throws Exception {
+        log.info("SpotlightController /oauth/refresh");
+
+        if (!roleService.isSuperAdmin(httpRequest)) {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java:126]: We may be able to use:

`@PreAuthorize("hasRole('SUPER_ADMIN')")`

File: src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java (hunk 1 above)
Comment (iaincooper-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java:135]: Fixed

File: src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java (hunk 2 above)
Comment (iaincooper-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java:178]: Fixed

File: src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java (hunk 3 above)
Comment (iaincooper-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/SpotlightService.java:47]: Fixed

File: src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java (hunk 4 above)
Comment (iaincooper-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java:123]: Fixed

File: src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java (hunk 5 above)
Comment (iaincooper-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java:157]: Fixed, added a dedicated SpotlightTokenException

File: src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java (hunk 6 above)
Comment (iaincooper-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/web/SpotlightController.java:126]: This didn't work, it resulted in the URL always being Access Denied. The existing implementation is consistent with the rest of the user service controllers, but I do agree as tech debt we should look at using those @PreAuthorize annotations to control access

**Review by dylanwrightCO — APPROVED** [review]

**Pull request merged**